RequestReplyExt is not implemented for ThrottlingRequest

I saw in your migration guide that you now had some nice extension for just writing `.reply_to(msg)`. Since I had to upgrade to a git version anyways to get axum 0.8 support I wanted to use that nice new syntax. Unfortunately,  the compiler is unhappy:

```
error[E0599]: no method named `reply_to` found for struct `ThrottlingRequest` in the current scope
  --> src/commands/config.rs:93:3
   |
89 | /     bot.send_message(
90 | |         msg.chat.id,
91 | |         display_config(&group)
92 | |     )
93 | |     .reply_to(msg)
   | |     -^^^^^^^^ method not found in `ThrottlingRequest<JsonRequest<SendMessage>>`
   | |_____|
   |
```

I believe that this is because the new trait `RequestReplyExt` is not implemented for `ThrottlingRequest<T: RequestReplyExt>`. I haven't looked further into the code but I hope it is possible to implement it?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -39,6 +39,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Now `InlineQueryResultsButton` serializes properly ([issue 1181](https://github.com/teloxide/teloxide/issues/1181))
 - Now `ThreadId` is able to serialize in multipart requests ([PR 1179](https://github.com/teloxide/teloxide/pull/1179))
 - Now stack does not overflow on dispatch ([issue 1154](https://github.com/teloxide/teloxide/issues/1154))
+- Implement `RequestReplyExt` and `RequestLinkPreviewExt` on setters from `teloxide_core::payloads` so syntax sugar can work on bot adaptors too ([PR 1270](https://github.com/teloxide/teloxide/pull/1270))
 
 ## 0.13.0 - 2024-08-16
 

--- a/crates/teloxide/src/sugar/request.rs
+++ b/crates/teloxide/src/sugar/request.rs
@@ -1,18 +1,16 @@
-//! Additions to [`JsonRequest`] and [`MultipartRequest`].
+//! Additions to multiple [`payloads`].
 //!
-//! [`JsonRequest`]: teloxide_core::requests::JsonRequest
-//! [`MultipartRequest`]: teloxide_core::requests::MultipartRequest
+//! [`payloads`]: teloxide_core::payloads
 
-use teloxide_core::{
-    payloads::*,
-    requests::{JsonRequest, MultipartRequest},
-    types::*,
-};
+use teloxide_core::{payloads::*, types::*};
 
 macro_rules! impl_request_reply_ext {
-    ($($t:ty),*) => {
+    ($(($t:ty, $trait:path)),*) => {
         $(
-            impl RequestReplyExt for $t {
+            impl<T> RequestReplyExt<$t> for T
+            where
+                T: $trait
+            {
                 fn reply_to<M>(self, message_id: M) -> Self
                 where
                     M: Into<MessageId>,
@@ -26,9 +24,12 @@ macro_rules! impl_request_reply_ext {
 }
 
 macro_rules! impl_request_link_preview_ext {
-    ($($t:ty),*) => {
+    ($(($t:ty, $trait:path)),*) => {
         $(
-            impl RequestLinkPreviewExt for $t {
+            impl<T> RequestLinkPreviewExt<$t> for T
+            where
+                T: $trait
+            {
                 fn disable_link_preview(self, is_disabled: bool) -> Self
                 where
                     Self: Sized
@@ -48,7 +49,7 @@ macro_rules! impl_request_link_preview_ext {
 }
 
 /// `.reply_to(msg)` syntax sugar for requests.
-pub trait RequestReplyExt {
+pub trait RequestReplyExt<P> {
     /// Replaces `.reply_parameters(ReplyParameters::new(msg.id))`
     /// with `.reply_to(msg.id)` or `.reply_to(msg)`
     fn reply_to<M>(self, message_id: M) -> Self
@@ -58,7 +59,7 @@ pub trait RequestReplyExt {
 }
 
 /// `.disable_link_preview(is_disabled)` syntax sugar for requests.
-pub trait RequestLinkPreviewExt {
+pub trait RequestLinkPreviewExt<P> {
     /// Replaces
     /// `.link_preview_options(LinkPreviewOptions {
     ///     is_disabled: true,
@@ -74,30 +75,37 @@ pub trait RequestLinkPreviewExt {
         Self: Sized;
 }
 
+// NOTE: struct in (struct, trait) is mandatory, as well as the blank <P>
+// generic in Request*Ext traits.
+// Or you will get `error[E0119]: conflicting implementations of trait`.
+// This is a Rust issue. See:
+// https://github.com/rust-lang/rust/issues/20400
+// https://github.com/rust-lang/rfcs/pull/1672
+
 impl_request_reply_ext! {
-    JsonRequest<SendDice>,
-    JsonRequest<SendInvoice>,
-    JsonRequest<SendPoll>,
-    JsonRequest<SendContact>,
-    JsonRequest<SendGame>,
-    JsonRequest<SendVenue>,
-    JsonRequest<SendLocation>,
-    JsonRequest<CopyMessage>,
-    JsonRequest<SendMessage>,
-    MultipartRequest<SendSticker>,
-    MultipartRequest<SendMediaGroup>,
-    MultipartRequest<SendAnimation>,
-    MultipartRequest<SendVideoNote>,
-    MultipartRequest<SendVideo>,
-    MultipartRequest<SendDocument>,
-    MultipartRequest<SendAudio>,
-    MultipartRequest<SendVoice>,
-    MultipartRequest<SendPhoto>
+    (SendDice, SendDiceSetters),
+    (SendInvoice, SendInvoiceSetters),
+    (SendPoll, SendPollSetters),
+    (SendContact, SendContactSetters),
+    (SendGame, SendGameSetters),
+    (SendVenue, SendVenueSetters),
+    (SendLocation, SendLocationSetters),
+    (CopyMessage, CopyMessageSetters),
+    (SendMessage, SendMessageSetters),
+    (SendSticker, SendStickerSetters),
+    (SendMediaGroup, SendMediaGroupSetters),
+    (SendAnimation, SendAnimationSetters),
+    (SendVideoNote, SendVideoNoteSetters),
+    (SendVideo, SendVideoSetters),
+    (SendDocument, SendDocumentSetters),
+    (SendAudio, SendAudioSetters),
+    (SendVoice, SendVoiceSetters),
+    (SendPhoto, SendPhotoSetters)
 }
 
 impl_request_link_preview_ext! {
-    JsonRequest<SendMessage>,
-    JsonRequest<EditMessageText>
+    (SendMessage, SendMessageSetters),
+    (EditMessageText, EditMessageTextSetters)
 }
 
 #[cfg(test)]
@@ -105,7 +113,7 @@ mod tests {
     use std::ops::Deref;
 
     use super::*;
-    use teloxide_core::{prelude::Requester, Bot};
+    use teloxide_core::{prelude::*, Bot};
 
     #[test]
     fn test_reply_to() {
@@ -136,6 +144,24 @@ mod tests {
     }
 
     #[test]
+    #[cfg(feature = "trace-adaptor")]
+    fn test_reply_to_adaptors() {
+        use teloxide_core::{adaptors::trace::Settings, requests::HasPayload};
+
+        let bot = Bot::new("TOKEN").trace(Settings::empty()).parse_mode(ParseMode::Html);
+
+        let real_reply_req = bot
+            .send_message(ChatId(1234), "test")
+            .reply_parameters(ReplyParameters::new(MessageId(1)));
+        let sugar_reply_req = bot.send_message(ChatId(1234), "test").reply_to(MessageId(1));
+
+        assert_eq!(
+            real_reply_req.payload_ref().payload_ref().reply_parameters,
+            sugar_reply_req.payload_ref().payload_ref().reply_parameters
+        )
+    }
+
+    #[test]
     fn test_disable_link_preview() {
         let link_preview_options = LinkPreviewOptions {
             is_disabled: true,
@@ -151,5 +177,29 @@ mod tests {
         let sugar_link_req = bot.send_message(ChatId(1234), "test").disable_link_preview(true);
 
         assert_eq!(real_link_req.deref(), sugar_link_req.deref())
+    }
+
+    #[test]
+    #[cfg(feature = "trace-adaptor")]
+    fn test_disable_link_preview_adaptors() {
+        use teloxide_core::{adaptors::trace::Settings, requests::HasPayload};
+
+        let link_preview_options = LinkPreviewOptions {
+            is_disabled: true,
+            url: None,
+            prefer_small_media: false,
+            prefer_large_media: false,
+            show_above_text: false,
+        };
+        let bot = Bot::new("TOKEN").trace(Settings::empty()).parse_mode(ParseMode::Html);
+
+        let real_link_req =
+            bot.send_message(ChatId(1234), "test").link_preview_options(link_preview_options);
+        let sugar_link_req = bot.send_message(ChatId(1234), "test").disable_link_preview(true);
+
+        assert_eq!(
+            real_link_req.payload_ref().payload_ref().reply_parameters,
+            sugar_link_req.payload_ref().payload_ref().reply_parameters
+        )
     }
 }

